masonry/src/widget/zstack.rs

```diff
@@ -0,0 +1,324 @@
+}
+
+impl From<(VerticalAlignment, HorizontalAlignment)> for Alignment {
+    fn from((vertical, horizontal): (VerticalAlignment, HorizontalAlignment)) -> Self {
+        Alignment::new(vertical, horizontal)
+    }
+}
+
+impl From<VerticalAlignment> for Alignment {
+    fn from(vertical: VerticalAlignment) -> Self {
+        Alignment::new(vertical, HorizontalAlignment::Center)
+    }
+}
+
+impl From<HorizontalAlignment> for Alignment {
+    fn from(horizontal: HorizontalAlignment) -> Self {
+        Alignment::new(VerticalAlignment::Center, horizontal)
+    }
+}
+
+// --- MARK: IMPL ZSTACK ---
+impl ZStack {
+    pub fn new() -> Self {
+        ZStack {
+            children: Vec::new(),
+            alignment: Alignment::default(),
+        }
+    }
+
+    pub fn with_alignment(mut self, alignment: impl Into<Alignment>) -> Self {
+        self.alignment = alignment.into();
+        self
+    }
+
+    pub fn with_child(self, child: impl Widget) -> Self {
+        self.with_child_pod(WidgetPod::new(Box::new(child)))
+    }
+
+    pub fn with_child_id(self, child: impl Widget, id: WidgetId) -> Self {
+        self.with_child_pod(WidgetPod::new_with_id(Box::new(child), id))
+    }
+
+    pub fn with_child_pod(mut self, child: WidgetPod<Box<dyn Widget>>) -> Self {
+        let child = Child { widget: child };
+        self.children.push(child);
+        self
+    }
+}
+
+// --- MARK: WIDGETMUT---
+impl ZStack {
+    pub fn add_child(this: &mut WidgetMut<'_, Self>, child: impl Widget) {
+        let child_pod: WidgetPod<Box<dyn Widget>> = WidgetPod::new(Box::new(child));
+        Self::insert_child_pod(this, child_pod);
+    }
+
+    pub fn add_child_id(this: &mut WidgetMut<'_, Self>, child: impl Widget, id: WidgetId) {
+        let child_pod: WidgetPod<Box<dyn Widget>> = WidgetPod::new_with_id(Box::new(child), id);
+        Self::insert_child_pod(this, child_pod);
+    }
+
+    pub fn insert_child_pod(this: &mut WidgetMut<'_, Self>, widget: WidgetPod<Box<dyn Widget>>) {
+        let child = Child { widget };
+        this.widget.children.push(child);
+        this.ctx.children_changed();
+        this.ctx.request_layout();
+    }
+
+    pub fn remove_child(this: &mut WidgetMut<'_, Self>, idx: usize) {
+        let child = this.widget.children.remove(idx);
+        this.ctx.remove_child(child.widget);
+        this.ctx.request_layout();
+    }
+
+    pub fn child_mut<'t>(
+        this: &'t mut WidgetMut<'_, Self>,
+        idx: usize,
+    ) -> Option<WidgetMut<'t, Box<dyn Widget>>> {
+        let child = &mut this.widget.children[idx].widget;
+        Some(this.ctx.get_mut(child))
+    }
+
+    pub fn set_alignment(this: &mut WidgetMut<'_, Self>, alignment: impl Into<Alignment>) {
+        this.widget.alignment = alignment.into();
+        this.ctx.request_layout();
+    }
+}
+
+// --- MARK: IMPL WIDGET---
+impl Widget for ZStack {
+    fn layout(&mut self, ctx: &mut LayoutCtx, bc: &BoxConstraints) -> Size {
+        let total_size = bc.max();
```

Comment:
This is suspect for a few reasons:
1) An infinite bound makes this fundamentally broken.
2) Even if all the children are small, this view takes up all available space.

I think the correct approach here might be a 2-pass solution, first to get the "overall" size, then to align widgets within that size. Allowing that is the reason that `run_layout` and `place_child` are different methods.

Reply:
I added this in [78a901e](https://github.com/linebender/xilem/pull/737/commits/78a901ec43ff76ca81752740eb2480b85c1a3591), can you please take a look at it. I'm unsure if I did it right?

Reply:
I replaced on of the `run_layout` calls with a `child_size` call instead.